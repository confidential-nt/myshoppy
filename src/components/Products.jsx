import React, { useEffect, useState } from "react";
import ProductCard from "./ProductCard";

export default function Products({ productRepository }) {
  const [products, setProducts] = useState();
  useEffect(() => {
    productRepository.findAll(setProducts);
  }, []);
  return (
    <ul className="flex mt-4 pr-2 pl-3">
      {products &&
        Object.entries(products).map(([key, value]) => (
          <ProductCard key={key} product={value} id={key} />
        ))}
    </ul>
  );
}
