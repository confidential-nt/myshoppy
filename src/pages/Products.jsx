import React from "react";

export default function Products() {
  return <div>여기는 products</div>;
}
