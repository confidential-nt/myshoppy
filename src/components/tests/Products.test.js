import { render, screen, waitFor } from "@testing-library/react";
import Products from "../Products";
import { withReactQuery, withRouter } from "../../tests/utils";
import { Route } from "react-router-dom";
import { fakeProducts } from "../../tests/products";

describe("Products", () => {
  const fakeProductRepository = {
    findAll: jest.fn(),
  };

  afterEach(() => fakeProductRepository.findAll.mockReset());

  it("스냅샷 테스트", async () => {
    fakeProductRepository.findAll.mockImplementation(() => fakeProducts);

    const { asFragment } = renderProducts();
    await screen.findAllByRole("listitem");
    expect(asFragment()).toMatchSnapshot();
  });

  it("로딩 중이면 로딩 표시가 떠야한다", () => {
    fakeProductRepository.findAll.mockImplementation(() => fakeProducts);
    renderProducts();
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("에러가 나면 에러가 났다는 표시가 떠야한다", async () => {
    fakeProductRepository.findAll.mockImplementation(() => {
      throw new Error();
    });
    renderProducts();
    await waitFor(() => {
      expect(screen.getByText("Something is wrong 😖")).toBeInTheDocument();
    });
  });

  it("products가 없으면 그 어떤 product card도 생성되지 말아야한다.", async () => {
    fakeProductRepository.findAll.mockImplementation(() => null);
    renderProducts();
    await waitFor(() => {
      expect(screen.queryByRole("listitem")).not.toBeInTheDocument();
    });
  });

  it("product 목록이 올바르게 렌더되어야한다", async () => {
    fakeProductRepository.findAll.mockImplementation(() => fakeProducts);
    renderProducts();
    await waitFor(() => {
      expect(screen.getAllByRole("listitem")).toHaveLength(
        Object.keys(fakeProducts).length
      );
    });
  });

  function renderProducts() {
    return render(
      withReactQuery(
        withRouter(
          <Route
            element={<Products productRepository={fakeProductRepository} />}
            path="/"
          />
        )
      )
    );
  }
});
