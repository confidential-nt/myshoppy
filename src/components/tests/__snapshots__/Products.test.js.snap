// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products 스냅샷 테스트 1`] = `
<DocumentFragment>
  <ul
    class="grid grid-cols-4 gap-3 gap-y-4 mt-4 pr-2 pl-3"
  >
    <li
      class="cursor-pointer shadow-lg hover:scale-105 transition"
    >
      <div
        class="w-full"
      >
        <img
          alt="핫핑크 후드 HOODY"
          class="w-full h-full"
          src="http://image/"
        />
      </div>
      <div
        class="p-4"
      >
        <div
          class="flex justify-between"
        >
          <h4>
            핫핑크 후드 HOODY
          </h4>
          <strong
            class="font-normal"
          >
            ₩40000
          </strong>
        </div>
        <span
          class="text-sm text-gray-700"
        >
          여성
        </span>
      </div>
    </li>
    <li
      class="cursor-pointer shadow-lg hover:scale-105 transition"
    >
      <div
        class="w-full"
      >
        <img
          alt="블랙 후드 HOODY"
          class="w-full h-full"
          src="http://image/"
        />
      </div>
      <div
        class="p-4"
      >
        <div
          class="flex justify-between"
        >
          <h4>
            블랙 후드 HOODY
          </h4>
          <strong
            class="font-normal"
          >
            ₩40000
          </strong>
        </div>
        <span
          class="text-sm text-gray-700"
        >
          여성
        </span>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
